woocommerce/woocommerce-ios
# [Bundle products in order] Move validation error to a notice at the bottom

Part of #11259
⚠️ Please review https://github.com/woocommerce/woocommerce-ios/pull/11292 first


## Why


From the design review w3JBReW1UJesCf02sj43kR-fi-984_9201, we'd like to move any validation error from either the top (bundle level) or inline (bundle item level) error message to a notice at the bottom.

The design looks like:
<img width="478" alt="Screenshot 2023-11-28 at 4 05 14 PM" src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/67f3aabe-f81d-4817-ae56-d52c86e72cae">

Please note that there are still some further enhancements we can make to match the design 100%:
- Update the bundle-level error message to match the design like `This bundle requires a total count of 2. Add 1 more of any product to continue.`. I personally found it a bit verbose and would like to get your thought on whether it's worth making this change
- Hide the notice when the bundle becomes valid, like after 1 second - I'll think about how to implement this

## How

A notice view is shown when:

- The bundle configuration is invalid
- The bundle configuration **becomes** valid
  - In order to implement this state change, I added `Publisher.withPrevious` publisher to observe the current & previous values of the validation state

In `ConfigurableBundleItemView` and `ConfigurableBundleProductView`, the error message text was removed. The new notice view is `ConfigurableBundleNoticeView`, which depends on a validation state `Result<Void, ValidationError>` from the view model `ConfigurableBundleProductViewModel`. I chose to only show the first error message so that the notice height doesn't take up a lot of vertical space when there are multiple errors.

For each item level validation error message in `ConfigurableBundleItemViewModel`, the product name is now included for the merchant to know which item to update.

## Testing instructions


Prerequisite: the store has [Product Bundles extension](https://woocommerce.com/products/product-bundles/) enabled, and also a bundle product **with a bundle size rule** and **at least 1 optional variable item** with non-empty variations and **at least 1 non-variable item**.

- Go to the Orders tab
- Tap `+` to create an order
- Tap `Add Products` and select the bundle product in the prerequisite --> a notice should be shown if the default quantities don't match the bundle size rule
- Configure the non-variable items so that they are valid --> the Save CTA should be enabled, and a green notice should be shown
- Tap the circle to select the variable item --> the Save CTA should be disabled while a notice with an error message appears
- Tap `Choose variation` to choose a variation
- In the configuration form, the attributes should be shown that look like the expected design
- Configure the variable item so that it's valid --> the Save CTA should be enabled, and a green notice should be shown
- Tap `Save Configuration`
- Tap `1 Product Selected` --> the bundle product should be added to the order with the expected configuration

## Screenshots


bundle level error | valid state - light mode | item level error | valid state - dark mode
-- | -- | -- | --
<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/93c930e7-a475-4f76-9d43-45b283beca09" width="300" /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/193fab1f-8c4a-4b42-8515-02a86d7ec8cf" width="300" /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/4fbe02a8-1e60-4863-8e03-088ed709bba5" width="300" /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/4e7d19b3-ad84-4275-ba7e-992ef34ad2d1" width="300" />

---
